File: java/Array&Hashing/ValidSudoku.java
import java.util.HashSet;

public class ValidSudoku {
    /**
     * 36. Valid Sudoku
     * Difficulty: Medium
     * Topics: Hash Table, Matrix
     */
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character>[] rows = new HashSet[9];
        HashSet<Character>[] cols = new HashSet[9];
        HashSet<Character>[] boxes = new HashSet[9];

        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<>();
            cols[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                char c = board[i][j];
                if (c == '.') continue;

                // Encode box constraints
                int boxIdx = (i / 3) * 3 + (j / 3);

                // If any key is already in the set, it's invalid
                if (rows[i].contains(c) || cols[j].contains(c) || boxes[boxIdx].contains(c)) {
                    return false;
                }

                // Otherwise, add keys to the set
                rows[i].add(c);
                cols[j].add(c);
                boxes[boxIdx].add(c);
            }
        }

        return true;
    }
}

File: python/Array&Hashing/ValidSudoku.py
"""
36. Valid Sudoku
Difficulty: Medium
Topics: Array, Hash Table, Matrix
"""
def isValidSudoku(self, board: list[list[str]]) -> bool:
    # We use 3 sets to keep track of what numbers we’ve already seen:
    row_check = [set() for _ in range(9)]
    col_check = [set() for _ in range(9)]
    box_check = [set() for _ in range(9)]

    for r in range(9):
        for c in range(9):
            if board[r][c] == ".":
                continue
            num = board[r][c]

            # Encode box constraints
            box_key = (r//3) * 3 + (c//3)

            # If any key is already in the set, it's invalid
            if num in row_check[r] or num in col_check[c] or num in box_check[box_key]:
                return False

            # Otherwise, add keys to the set
            row_check[r].add(num)
            col_check[c].add(num)
            box_check[box_key].add(num)

    return True